src/core/cpu_patches.cpp

```diff
@@ -0,0 +1,381 @@
+    const auto dst = ZydisToXbyakRegisterOperand(operands[0]);
+    const auto src = ZydisToXbyakOperand(operands[1]);
+    const auto start_len = ZydisToXbyakRegisterOperand(operands[2]);
+
+    const Xbyak::Reg32e shift(Xbyak::Operand::RCX, static_cast<int>(start_len.getBit()));
+    const auto scratch1 =
+        AllocateScratchRegister({&dst, src.get(), &start_len, &shift}, dst.getBit());
+    const auto scratch2 =
+        AllocateScratchRegister({&dst, src.get(), &start_len, &shift, &scratch1}, dst.getBit());
+
+    if (dst.getIdx() == shift.getIdx()) {
+        SaveRegisters(c, {scratch1, scratch2});
+    } else {
+        SaveRegisters(c, {scratch1, scratch2, shift});
+    }
+
+    c.mov(scratch1, *src);
+    if (shift.getIdx() != start_len.getIdx()) {
+        c.mov(shift, start_len);
+    }
+
+    c.shr(scratch1, shift.cvt8());
+    c.shr(shift, 8);
+    c.mov(scratch2, 1);
+    c.shl(scratch2, shift.cvt8());
+    c.dec(scratch2);
+
+    c.mov(dst, scratch1);
+    c.and_(dst, scratch2);
+
+    if (dst.getIdx() == shift.getIdx()) {
+        RestoreRegisters(c, {scratch1, scratch2});
+    } else {
+        RestoreRegisters(c, {scratch1, scratch2, shift});
+    }
+
+    return true;
+}
+
+static bool GenerateBLSI(const ZydisDecodedOperand* operands, Xbyak::CodeGenerator& c) {
+    const auto dst = ZydisToXbyakRegisterOperand(operands[0]);
+    const auto src = ZydisToXbyakOperand(operands[1]);
+
+    const auto scratch = AllocateScratchRegister({&dst, src.get()}, dst.getBit());
+
+    SaveRegisters(c, {scratch});
+
+    c.mov(scratch, *src);
+    c.neg(scratch);
+    c.and_(scratch, *src);
+    c.mov(dst, scratch);
+
+    RestoreRegisters(c, {scratch});
+
+    return true;
+}
+
+static bool GenerateBLSMSK(const ZydisDecodedOperand* operands, Xbyak::CodeGenerator& c) {
+    const auto dst = ZydisToXbyakRegisterOperand(operands[0]);
+    const auto src = ZydisToXbyakOperand(operands[1]);
+
+    const auto scratch = AllocateScratchRegister({&dst, src.get()}, dst.getBit());
+
+    SaveRegisters(c, {scratch});
+
+    c.mov(scratch, *src);
+    c.dec(scratch);
+    c.xor_(scratch, *src);
+    c.mov(dst, scratch);
+
+    RestoreRegisters(c, {scratch});
+
+    return true;
+}
+
+static bool GenerateBLSR(const ZydisDecodedOperand* operands, Xbyak::CodeGenerator& c) {
+    const auto dst = ZydisToXbyakRegisterOperand(operands[0]);
+    const auto src = ZydisToXbyakOperand(operands[1]);
+
+    const auto scratch = AllocateScratchRegister({&dst, src.get()}, dst.getBit());
+
+    SaveRegisters(c, {scratch});
+
+    c.mov(scratch, *src);
+    c.dec(scratch);
+    c.and_(scratch, *src);
+    c.mov(dst, scratch);
+
+    RestoreRegisters(c, {scratch});
+
+    return true;
+}
+
+#endif // __APPLE__
+
+using InstructionGenerator = bool (*)(const ZydisDecodedOperand*, Xbyak::CodeGenerator&);
+static const std::unordered_map<ZydisMnemonic, InstructionGenerator> InstructionGenerators = {
+    {ZYDIS_MNEMONIC_MOV, GenerateTlsAccess},
+#ifdef __APPLE__
```

Comment:
i wonder if that can be more generic , for example using xbyak cpu feature detection , so it can work on other platforms that doesn't support BMI1 . Also same stuff should be done for SSE4a which is amd only and it isn't supported in rosseta 2 too (i assume)

Reply:
It's possible to use feature detection but I don't think it would be very useful. It's a unique situation for Rosetta 2 that it supports newer extensions like AVX but not all of BMI1; you probably wouldn't find such a situation anywhere else. And supporting patching out AVX and others as well would add a lot of complexity.

For SSE4a I'm not sure on the details, but I'm mainly bringing it up to parity with the other platforms here. If that's an issue running the emulator on e.g. Windows with an Intel CPU, could be something to explore down the line.

Reply:
SSE4a is amd exclusive (only 4 instructions) . That is an issue for all platforms that has intel cpu , and since it is so rare i am sure also rosseta doesn't have it

Reply:
With this new patch framework I feel we can add any instructions we see relatively easily yeah